File: array/medium/longest_subarray_sum_k.py
# Given an array and a sum k, we need to print the length of the longest subarray that sums to k.

def longest_subarray_sum_k(array, k):
    longest_subarray = 0
    for i in range(len(array)):
        long = 0
        if (array[i] == k):
            long += 1
            if (long > longest_subarray):
                longest_subarray = long
            continue
        sum = 0
        for j in range(i, len(array)):
            sum += array[j]
            if (sum == k):
                long = (j - i) + 1
                # longest_subarray += 1
                break
            if (sum > k):
                break
        if (long > longest_subarray):
            longest_subarray = long
    return longest_subarray

if __name__=="__main__":
    array = [1,-1,0]
    print(longest_subarray_sum_k(array, 0))

File: array/medium/longest_subarray_sum_k.js
//Given an array and a sum k, we need to print the length of the longest subarray that sums to k.(all positives)

const longestSubarray = (array, k) => {
  let left = 0;
  let right = 0;
  let sum = array[0];
  let maxLen = 0;
  while (right < array.length) {
    if (left <= right && sum > k) {
      sum -= array[left];
      left++;
    }

    if (sum === k) {
      maxLen = Math.max(maxLen, right - left + 1);
    }

    right++;
    sum += array[right];
  }
  return maxLen;
};

console.log(longestSubarray([2, 3, 5, 1, 9], 10));
